# Impact of Prompting Strategies on Information Extraction from AI News Articles Using Open-Source LLMs

## Part 1: Data Preparation and Exploratory Data Analysis (EDA)

In [1]:
import pandas as pd

# Import dataset
df = pd.read_csv("MIT_AI_ARTICLES.csv")

print("Original article number:", len(df))

# Hold on only necessary columns
df = df[[
    "title",
    "publication_date",
    "summary",
    "body",
    "url"
]]

# Remove the ones with an empty Body.
df = df.dropna(subset=["body"])

# Calculate word count
df["word_count"] = df["body"].astype(str).str.split().str.len()

# Remove shorter or longer articles
clean_df = df[
    (df["word_count"] >= 200) &
    (df["word_count"] <= 3000)
]

print("Number of cleaned articles:", len(clean_df))

# Save
clean_df.to_csv("clean_articles.csv", index=False)

clean_df.head()

Original article number: 314
Number of cleaned articles: 314


,title,publication_date,summary,body,url,word_count
0,A new model predicts how molecules will dissol...,"August 19, 2025",Solubility predictions could make it easier to...,"Using machine learning, MIT chemical engineers...",https://news.mit.edu/2025/new-model-predicts-h...,1149
1,Researchers glimpse the inner workings of prot...,"August 18, 2025",A new approach can reveal the features AI mode...,"Within the past few years, models that can pre...",https://news.mit.edu/2025/researchers-glimpse-...,1025
2,How AI could speed the development of RNA vacc...,"August 15, 2025",MIT engineers used a machine-learning model to...,"Using artificial intelligence, MIT researchers...",https://news.mit.edu/2025/how-ai-could-speed-d...,1006
3,"Using generative AI, researchers design compou...","August 14, 2025",The team used two different AI approaches to d...,"With help from artificial intelligence, MIT re...",https://news.mit.edu/2025/using-generative-ai-...,1076
4,A new way to test how well AI systems classify...,"August 13, 2025",As large language models increasingly dominate...,Is this movie review a rave or a pan? Is this ...,https://news.mit.edu/2025/new-way-test-how-wel...,1186


In [2]:
clean_df["word_count"].describe()

count     314.000000
mean     1053.404459
std       283.800000
min       259.000000
25%       910.750000
50%      1051.500000
75%      1172.000000
max      2304.000000
Name: word_count, dtype: float64

After removing missing values and filtering articles outside the 200–3000 word range, all 314 articles remained in the dataset. 
Therefore, no additional article removal was required.

#### Control of numbers company, model and tech at all dataset

In [3]:
import re
from collections import Counter

companies = [
    "OpenAI",
    "Google",
    "Meta",
    "Microsoft",
    "Anthropic",
    "Nvidia",
    "Amazon",
    "IBM",
    "Adobe",
    "Tesla",
    "Apple",
    "DeepMind"
]

all_text = " ".join(clean_df["body"].astype(str))

for company in companies:
    count = len(re.findall(company, all_text, flags=re.IGNORECASE))
    print(f"{company}: {count}")

OpenAI: 28
Google: 79
Meta: 108
Microsoft: 14
Anthropic: 6


Nvidia: 21
Amazon: 34
IBM: 107
Adobe: 8
Tesla: 4
Apple: 17
DeepMind: 9


In [4]:
models = [
    "GPT",
    "Claude",
    "Gemini",
    "Llama",
    "Mistral"
]

for model in models:
    print(model, all_text.count(model))

GPT 115
Claude 13
Gemini 5
Llama 4
Mistral 3


In [5]:
techs = [
    "Generative AI",
    "Large Language Model",
    "Machine Learning",
    "Deep Learning",
    "Reinforcement Learning",
    "Neural Network",
    "Computer Vision",
    "Multimodal"
]

for tech in techs:
    print(tech, all_text.count(tech))

Generative AI 24
Large Language Model 0
Machine Learning 42
Deep Learning 2
Reinforcement Learning 0
Neural Network 0
Computer Vision 15
Multimodal 2


## Part 2 : Selecting 100 articles

In [6]:
keywords = [
    "OpenAI",
    "Google",
    "Meta",
    "Microsoft",
    "Anthropic",
    "Nvidia",
    "GPT",
    "Claude",
    "Gemini",
    "Llama"
]

mask = clean_df["body"].astype(str).apply(
    lambda x: any(k.lower() in x.lower() for k in keywords)
)
entity_articles = clean_df[mask]

eval_df = entity_articles.sample(n=100, random_state=42).reset_index(drop=True)

eval_df.insert(0, "article_id", [f"art_{i:03d}" for i in range(len(eval_df))])

eval_df.to_csv("evaluation_articles.csv", index=False)

print("Number of articles:", len(eval_df))
eval_df[["article_id", "title", "word_count"]].head()

Number of articles: 100


,article_id,title,word_count
0,art_000,Stratospheric safety standards: How aviation c...,1513
1,art_001,Can robots learn from machine dreams?,1059
2,art_002,School of Engineering welcomes new faculty,1564
3,art_003,Envisioning a future where health care tech le...,1084
4,art_004,Introducing the MIT Generative AI Impact Conso...,1652


## Part 3 — Building the Annotation Sheet (Ground Truth)

In [7]:
CANDIDATE_COMPANIES = [
    "OpenAI", "Anthropic", "Google DeepMind", "Google", "DeepMind",
    "Meta", "Microsoft", "Nvidia", "Amazon", "IBM", "Adobe", "Tesla", "Apple",
]
CANDIDATE_MODELS = [
    "GPT-5", "GPT-4", "GPT", "Claude", "Gemini", "Llama", "Mistral",
]
TECHNOLOGY_LABELS = [
    "Retrieval-Augmented Generation", "RAG",
    "Reinforcement Learning",
    "Generative AI",
    "Multimodal AI",
    "AI Agents",
]

def find_mentions(text, candidates):
    text_lower = str(text).lower()
    found = []
    for candidate in candidates:
        if candidate.lower() in text_lower:
            found.append(candidate)
    return found

annotation_rows = []
for _, row in eval_df.iterrows():
    suggested_companies = find_mentions(row["body"], CANDIDATE_COMPANIES)
    suggested_models = find_mentions(row["body"], CANDIDATE_MODELS)
    suggested_technologies = find_mentions(row["body"], TECHNOLOGY_LABELS)

    annotation_rows.append({
        "article_id": row["article_id"],
        "title": row["title"],
        "url": row["url"],
        "body": row["body"],
        # Pre-filled guesses 
        "suggested_companies": "; ".join(suggested_companies),
        "suggested_models": "; ".join(suggested_models),
        "suggested_technologies": "; ".join(suggested_technologies),
        # I fill these in by hand with the CORRECT, final entity list.
        "gt_companies": "",
        "gt_models": "",
        "gt_technologies": "",
    })

annotation_df = pd.DataFrame(annotation_rows)
annotation_df.to_csv("annotation_sheet.csv", index=False)

print("annotation_sheet.csv created:", len(annotation_df), "rows")

annotation_df.drop(columns=["body"]).head()

annotation_sheet.csv created: 100 rows


,article_id,title,url,suggested_companies,suggested_models,suggested_technologies,gt_companies,gt_models,gt_technologies
0,art_000,Stratospheric safety standards: How aviation c...,https://news.mit.edu/2024/stratospheric-safety...,Microsoft,,RAG,,,
1,art_001,Can robots learn from machine dreams?,https://news.mit.edu/2024/can-robots-learn-mac...,Amazon,GPT,Generative AI,,,
2,art_002,School of Engineering welcomes new faculty,https://news.mit.edu/2024/school-engineering-w...,Google; DeepMind; Meta,,RAG; Reinforcement Learning,,,
3,art_003,Envisioning a future where health care tech le...,https://news.mit.edu/2025/envisioning-future-w...,Meta,,RAG,,,
4,art_004,Introducing the MIT Generative AI Impact Conso...,https://news.mit.edu/2025/introducing-mit-gene...,OpenAI,GPT,RAG; Generative AI; AI Agents,,,


## Part 4: Prompt Design

Three prompting strategies were implemented in `prompt_templates.py`:
- **Zero-shot**: task description only.
- **Few-shot**: task description plus two annotated example articles.
- **Structured**: task description plus an explicit JSON output schema.

AI technology extraction is restricted to the five predefined labels used during ground truth annotation. Companies and models remain open-set.

In [8]:
from prompt_templates import build_prompt, PROMPT_BUILDERS

# I use a real evaluation article to preview the prompts
eval_df = pd.read_csv("evaluation_articles.csv")
sample_article = eval_df.loc[eval_df["article_id"] == "art_004", "body"].values[0]

print(build_prompt("zero_shot", sample_article)[:800])

Extract all company names, AI model names, and AI technologies mentioned in the article.
Only consider the following AI technologies (use these exact labels, and only include a technology if it is explicitly mentioned in the article): Retrieval-Augmented Generation (RAG), Reinforcement Learning, Generative AI, Multimodal AI, AI Agents.

Report the results as three labeled lists, in this exact format:
Companies:
- ...
AI Models:
- ...
AI Technologies:
- ...

If a category has no entities, leave its list empty.

Article:
From crafting complex code to revolutionizing the hiring process, generative artificial intelligence is reshaping industries faster than ever before — pushing the boundaries of creativity, productivity, and collaboration across countless domains.
Enter theMIT Generative AI I


In [9]:
print(build_prompt("few_shot", sample_article)[:1200])

Extract all company names, AI model names, and AI technologies mentioned in the article.
Only consider the following AI technologies (use these exact labels, and only include a technology if it is explicitly mentioned in the article): Retrieval-Augmented Generation (RAG), Reinforcement Learning, Generative AI, Multimodal AI, AI Agents.

Report the results as three labeled lists, in the same format as the examples below. If a category has no entities, leave its list empty.

Example 1
Article:
OpenAI announced GPT-5 while introducing new Retrieval-Augmented Generation capabilities. Anthropic also reported improvements to Claude.

Output:
Companies:
- OpenAI
- Anthropic
AI Models:
- GPT-5
- Claude
AI Technologies:
- Retrieval-Augmented Generation (RAG)

Example 2
Article:
Google DeepMind unveiled Gemini's new multimodal reasoning capabilities, allowing the model to process images, audio, and text together. Meta's research team separately published work on reinforcement learning agents tha

In [10]:
print(build_prompt("structured", sample_article)[:800])

Extract all company names, AI model names, and AI technologies mentioned in the article.
Only consider the following AI technologies (use these exact labels, and only include a technology if it is explicitly mentioned in the article): Retrieval-Augmented Generation (RAG), Reinforcement Learning, Generative AI, Multimodal AI, AI Agents.

Return ONLY a JSON object with this exact structure, and nothing else (no explanations, no markdown code fences):
{
  "companies": [],
  "models": [],
  "technologies": []
}

If a category has no entities, return an empty list for it.

Article:
From crafting complex code to revolutionizing the hiring process, generative artificial intelligence is reshaping industries faster than ever before — pushing the boundaries of creativity, productivity, and collabora


All three prompt builders take an article body and return a ready-to-send prompt string. These will be used in Part 5 (Model Inference) to query Llama 3 and Gemma 3 across the six model x strategy combinations.

## Part 5: Model Inference

Inference was run separately via `run_inference.py` rather than inline in this notebook, since it issues real API calls (with retries and incremental checkpointing) and takes several minutes to complete.

**Setup**
- Provider: OpenRouter (OpenAI-compatible API)
- Models: `meta-llama/llama-3.1-8b-instruct` (closest currently-available match to "Llama 3", since the original Llama 3 release has been deprecated on OpenRouter) and `google/gemma-3-12b-it`
- 100 evaluation articles x 3 prompting strategies x 2 models = 600 API calls
- `temperature=0.0` for deterministic, reproducible extractions

The script is resumable: it skips any (article_id, model, strategy) triple already present in `raw_extractions.csv`, so an interrupted run can simply be re-launched without repeating (and re-paying for) completed calls.

In [11]:
raw_df = pd.read_csv("raw_extractions.csv")

print("Total rows:", len(raw_df))
print("Unique articles:", raw_df["article_id"].nunique())
print("Rows with an error:", raw_df["error"].notna().sum())
print()
print(raw_df.groupby(["model", "strategy"]).size())

Total rows: 600
Unique articles: 100
Rows with an error: 0

model   strategy  
gemma3  few_shot      100
        structured    100
        zero_shot     100
llama3  few_shot      100
        structured    100
        zero_shot     100
dtype: int64


All 600 combinations completed successfully with no errors and no empty responses.

In [12]:
# Preview one structured-prompting output
sample = raw_df[
    (raw_df["article_id"] == "art_004")
    & (raw_df["model"] == "gemma3")
    & (raw_df["strategy"] == "structured")
].iloc[0]

print(sample["raw_output"])

```json
{
  "companies": [
    "Analog Devices",
    "The Coca-Cola Co.",
    "OpenAI",
    "Tata Group",
    "SK Telecom",
    "TWG Global",
    "Tata Sons Ltd."
  ],
  "models": [
    "ChatGPT",
    "UNIVAC"
  ],
  "technologies": [
    "Generative AI",
    "Retrieval-Augmented Generation (RAG)",
    "Reinforcement Learning",
    "Multimodal AI",
    "AI Agents"
  ]
}
```


**Observation.** Manual inspection of a few outputs during a smoke test (3 articles) surfaced a recurring failure pattern: for `art_002`, an article about new faculty hires, Gemma 3 listed universities (MIT, Stanford, Harvard, ...) as "companies", and under the structured prompt it returned all five predefined technology labels even though most were not explicitly mentioned in the article. This kind of over-generation / hallucination is exactly what Part 8 (Error Analysis) will quantify and compare across models and prompting strategies — it was left as-is rather than patched, since it is part of the behavior being measured.

## Part 6: Output Parsing

Raw model responses come in two formats and need different parsing logic, implemented in `parse_outputs.py`:
- **structured**: a JSON object, sometimes wrapped in a ` ```json ` fence, sometimes preceded by commentary text with no fence at all.
- **zero-shot / few-shot**: free text with `Companies:` / `AI Models:` / `AI Technologies:` section headers (with inconsistent markdown, e.g. `**Companies:**`) followed by bullet lists.

In [13]:
parsed_df = pd.read_csv("parsed_extractions.csv")

print("Total rows:", len(parsed_df))
print("Parse failures:", (~parsed_df["parse_success"]).sum())
print()
print(parsed_df.groupby(["model", "strategy"]).size())

Total rows: 600
Parse failures: 3

model   strategy  
gemma3  few_shot      100
        structured    100
        zero_shot     100
llama3  few_shot      100
        structured    100
        zero_shot     100
dtype: int64


**Two parsing issues were found and fixed while building this step (not just written correctly on the first try):**

1. **Header format mismatch.** The first version of the header regex only matched `**Companies**:` (colon after the closing bold markers), but Llama 3's actual output used `**Companies:**` (colon inside the bold markers). This alone caused 163/600 rows (27%) to fail parsing — almost entirely Llama 3 zero-shot and few-shot responses. Fixing the regex to accept the colon on either side of the markdown brought failures down to 3/600 (0.5%), all genuine malformed JSON from Llama 3's structured responses (e.g. an unbalanced parenthesis inside a string).

2. **Self-negating list items.** Llama 3 frequently writes bullet items that argue *against* including an entity instead of simply omitting it, e.g. `"Retrieval-Augmented Generation (RAG) (not mentioned in the article)"`. Left as-is, these would have been parsed as positive predictions and inflated false positives in the precision/recall evaluation (Step 7) — the opposite of what the model was actually saying. 143 such items were found across the dataset (mostly in the technologies field) and are now filtered out during parsing.

The remaining 3 failures are left as recorded failures (not patched) since they reflect a real JSON-formatting error by the model, which is exactly the kind of error Step 8 is meant to quantify.

In [14]:
parsed_df[~parsed_df["parse_success"]][["article_id", "model", "strategy", "parse_error"]]

,article_id,model,strategy,parse_error
50,art_008,llama3,structured,"invalid_json: Expecting ',' delimiter: line 4 ..."
80,art_013,llama3,structured,"invalid_json: Expecting ',' delimiter: line 4 ..."
152,art_025,llama3,structured,"invalid_json: Expecting ',' delimiter: line 4 ..."


## Part 7: Performance Evaluation

`evaluate.py` compares `parsed_extractions.csv` against `ground_truth_annotations.csv` and computes Precision, Recall, and F1-score, reported separately for companies, models, and technologies, plus an overall row per (model, strategy) that combines all three categories. Metrics are **micro-averaged**: TP/FP/FN counts are summed across all 100 articles for a given (model, strategy, category) before computing precision/recall/F1.

Entities are matched after normalization (lowercase, parenthetical asides stripped, punctuation/whitespace collapsed). This resolves a label-format mismatch between the ground truth ("Retrieval-Augmented Generation") and the prompt-instructed label ("Retrieval-Augmented Generation (RAG)") without a hardcoded special case, and also lets predictions like "MIT (Massachusetts Institute of Technology)" match the ground truth's "MIT". It does **not** resolve genuine naming variation (e.g. "Google DeepMind" vs. "DeepMind" as separate strings) — those remain real precision/recall errors, which is what Step 8 will look at more closely.

Command used:
```
python evaluate.py
```

In [15]:
metrics_df = pd.read_csv("evaluation_metrics.csv")

print("Overall (all three categories combined):")
metrics_df[metrics_df["category"] == "overall"].sort_values("f1", ascending=False)

Overall (all three categories combined):


,model,strategy,category,tp,fp,fn,precision,recall,f1
15,llama3,few_shot,overall,312,504,86,0.3824,0.7839,0.5140
3,gemma3,few_shot,overall,347,685,51,0.3362,0.8719,0.4853
11,gemma3,zero_shot,overall,353,735,45,0.3244,0.8869,0.4751
19,llama3,structured,overall,343,707,55,0.3267,0.8618,0.4738
7,gemma3,structured,overall,351,766,47,0.3142,0.8819,0.4634
23,llama3,zero_shot,overall,351,773,47,0.3123,0.8819,0.4612


In [16]:
for cat in ["companies", "models", "technologies"]:
    print(f"--- {cat} ---")
    display(metrics_df[metrics_df["category"] == cat].sort_values("f1", ascending=False))

--- companies ---


,model,strategy,category,tp,fp,fn,precision,recall,f1
16,llama3,structured,companies,154,276,32,0.3581,0.8280,0.5000
20,llama3,zero_shot,companies,161,346,25,0.3176,0.8656,0.4646
12,llama3,few_shot,companies,159,377,27,0.2966,0.8548,0.4404
0,gemma3,few_shot,companies,163,427,23,0.2763,0.8763,0.4201
8,gemma3,zero_shot,companies,166,441,20,0.2735,0.8925,0.4187
4,gemma3,structured,companies,163,445,23,0.2681,0.8763,0.4106


--- models ---


,model,strategy,category,tp,fp,fn,precision,recall,f1
13,llama3,few_shot,models,108,31,25,0.7770,0.8120,0.7941
21,llama3,zero_shot,models,111,48,22,0.6981,0.8346,0.7603
17,llama3,structured,models,116,59,17,0.6629,0.8722,0.7532
1,gemma3,few_shot,models,110,87,23,0.5584,0.8271,0.6667
5,gemma3,structured,models,114,97,19,0.5403,0.8571,0.6628
9,gemma3,zero_shot,models,113,101,20,0.5280,0.8496,0.6513


--- technologies ---


,model,strategy,category,tp,fp,fn,precision,recall,f1
2,gemma3,few_shot,technologies,74,171,5,0.3020,0.9367,0.4568
10,gemma3,zero_shot,technologies,74,193,5,0.2772,0.9367,0.4277
14,llama3,few_shot,technologies,45,96,34,0.3191,0.5696,0.4091
6,gemma3,structured,technologies,74,224,5,0.2483,0.9367,0.3926
22,llama3,zero_shot,technologies,79,379,0,0.1725,1.0000,0.2942
18,llama3,structured,technologies,73,372,6,0.1640,0.9241,0.2786


**Observations.**

- Recall is consistently high (0.78–0.89 overall) but precision is low (0.31–0.38 overall) across every model/strategy combination. This matches the over-generation behavior noticed during Part 5/6: both models tend to list many more entities than the article actually mentions — for companies specifically, this is largely driven by universities and research institutes (MIT, Stanford, Harvard, ...) being labeled as "companies".
- **Models** (AI model names) is the strongest category by far (F1 up to 0.79 for Llama 3 few-shot), likely because model names like "ChatGPT" or "GPT-4" are lexically distinctive and rarely confused with something else.
- **Technologies** shows the widest spread between strategies for Llama 3: zero-shot recall is a perfect 1.0 but precision is only 0.17, consistent with the pattern already seen in Part 5/6 of Llama 3 simply listing all five predefined technology labels regardless of whether they are mentioned. Few-shot prompting clearly helps Llama 3 here (F1 0.41 vs. 0.29 zero-shot), suggesting the examples teach it to be more selective — Gemma 3 does not show the same gain, its three strategies score similarly on technologies.
- Across strategies, **few-shot** gives the best overall F1 for both models, while **structured** and **zero-shot** trail slightly. This will be examined further in Part 8.

## Part 8: Error Analysis

`error_analysis.py` re-uses the same normalization and ground-truth comparison as `evaluate.py` (Part 7), but instead of aggregating straight to Precision/Recall/F1, it records one row per individual error instance, classified into four types:

- **missed_entity** — a ground-truth entity the model did not predict
- **entity_confusion** — a predicted entity that IS in the ground truth, but filed under a different category (e.g. a model name predicted as a company)
- **false_positive_entity** — a predicted entity not in the ground truth under any category (hallucinated, or genuinely not mentioned)
- **json_formatting_failure** — a structured-prompt response that failed to parse (carried over from Part 6)

As a sanity check, the missed_entity and (false_positive_entity + entity_confusion) counts here were verified to sum to exactly the FN and FP totals reported in `evaluation_metrics.csv` from Part 7, so the two steps agree with each other.

Command used:
```
python error_analysis.py
```

In [17]:
error_df = pd.read_csv("error_analysis.csv")

print("Total error instances:", len(error_df))
error_df.groupby(["model", "strategy", "error_type"]).size().unstack(fill_value=0)

Total error instances: 4504


error_type         entity_confusion  false_positive_entity  \
model  strategy                                              
gemma3 few_shot                   1                    684   
       structured                 2                    764   
       zero_shot                  2                    733   
llama3 few_shot                   1                    503   
       structured                 5                    702   
       zero_shot                  0                    773   

error_type         json_formatting_failure  missed_entity  
model  strategy                                            
gemma3 few_shot                          0             51  
       structured                        0             47  
       zero_shot                         0             45  
llama3 few_shot                          0             86  
       structured                        3             55  
       zero_shot                         0             47

### Finding 1 — "MIT" dominates the false positives

The single most common false-positive company, by far, is **"MIT"** itself: it was predicted as a company in 98 of the 100 articles, across every model and strategy (477 of the 2,306 false-positive company instances, i.e. roughly 1 in 5). This makes sense once you look at the data: the dataset is scraped from MIT News, so almost every article opens with something like "MIT researchers developed..." or "MIT engineers designed...". The ground truth annotation deliberately does **not** count MIT as an "AI company" in this context, since these are MIT's own news articles about MIT-affiliated research, not reporting on MIT as a commercial AI company the way an article would discuss OpenAI or Anthropic. Both models have no way to know this distinction from the prompt alone, so they extract MIT as a company essentially every time it appears syntactically like one.

In [18]:
fp_companies = error_df[
    (error_df["error_type"] == "false_positive_entity")
    & (error_df["category"] == "companies")
]
fp_companies["entity"].value_counts().head(10)

entity
mit                               477
harvard university                 59
ibm                                46
stanford university                35
carnegie mellon university         28
cornell university                 27
mit-ibm watson ai lab              24
national science foundation        24
mit sloan school of management     23
mit lincoln laboratory             18
Name: count, dtype: int64

### Finding 2 — Entity confusion is rare, but consistent

Only 11 of the 4,504 recorded errors (0.2%) are true entity confusion (the entity is real and in the ground truth, just filed under the wrong category). The pattern is always the same kind of mix-up: AI **product names** (ChatGPT, Midjourney, AlphaFold) get predicted as **companies**, and "Generative AI" — a technology label — gets predicted as a **model**. This is an understandable mistake: product names often look and read like company names, and "Generative AI" is phrased more like a product category than the other technology labels.

In [19]:
error_df[error_df["error_type"] == "entity_confusion"][
    ["model", "strategy", "article_id", "category", "entity", "detail"]
]

,model,strategy,article_id,category,entity,detail
428,llama3,structured,art_005,models,generative ai,ground truth lists it under 'technologies'
614,llama3,structured,art_010,companies,alphafold,ground truth lists it under 'models'
1857,llama3,structured,art_038,models,generative ai,ground truth lists it under 'technologies'
2153,gemma3,structured,art_044,companies,chatgpt,ground truth lists it under 'models'
2915,llama3,structured,art_066,models,generative ai,ground truth lists it under 'technologies'
3040,llama3,few_shot,art_069,companies,midjourney,ground truth lists it under 'models'
3051,llama3,structured,art_069,models,openai,ground truth lists it under 'companies'
3057,gemma3,zero_shot,art_069,companies,midjourney,ground truth lists it under 'models'
3060,gemma3,zero_shot,art_069,models,generative ai,ground truth lists it under 'technologies'
3064,gemma3,few_shot,art_069,companies,midjourney,ground truth lists it under 'models'


### Finding 3 — JSON formatting failures

All 3 JSON formatting failures happened with Llama 3 under the structured prompt, and all 3 fail at the exact same kind of spot: `Expecting ',' delimiter` a few characters into the technologies array. Looking at the raw output (Part 6), the cause is the same every time — Llama 3 writes `"Retrieval-Augmented Generation (RAG"` with the closing parenthesis placed *after* the closing quote instead of before it, breaking the JSON string. Gemma 3 never made this mistake in this dataset.

In [20]:
error_df[error_df["error_type"] == "json_formatting_failure"][
    ["model", "strategy", "article_id", "detail"]
]

,model,strategy,article_id,detail
526,llama3,structured,art_008,"invalid_json: Expecting ',' delimiter: line 4 ..."
716,llama3,structured,art_013,"invalid_json: Expecting ',' delimiter: line 4 ..."
1297,llama3,structured,art_025,"invalid_json: Expecting ',' delimiter: line 4 ..."


### Summary

Combining Parts 7 and 8: the low precision seen in the evaluation results is overwhelmingly driven by **false positives**, not entity confusion or formatting failures — genuine category mix-ups are vanishingly rare (11 instances) and JSON errors are a Llama 3 / structured-prompt-specific edge case (3 instances). The dominant error mode is over-generation: both models extract entities that are syntactically company/technology-like but are either not actually the article's subject (MIT) or not explicitly present in the text at all (the technology labels).